Add simple data-race detector

JCTyblaidd opened the pull request:
Partially fixes data-race detection, see #1372, based on Dynamic Race Detection for C++11

- This does not explore weak memory behaviour, only exploring one sequentially consistent ordering.
- Data-race detection is only enabled after the first thread is created, so should have minimal overhead for non-concurrent execution.
- ~~Does not attempt to re-use thread id's so creating and joining threads lots of time in an execution will result in the vector clocks growing in size and slowing down program execution~~ It does now

o752d commented on src/data_race.rs:
```diff
@@ -0,0 +1,1407 @@
+
+use smallvec::SmallVec;
+
+use crate::*;
```
```suggestion
use crate::MiriEvalContext;
```
Since there are only a couple of includes using `crate` below?

o752d commented on src/data_race.rs:
```diff
@@ -0,0 +1,1407 @@
+}
+
+/// Evaluation context extensions
+impl<'mir, 'tcx: 'mir> EvalContextExt<'mir, 'tcx> for crate::MiriEvalContext<'mir, 'tcx> {}
+pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx> {
```
```suggestion
impl<'mir, 'tcx: 'mir> EvalContextExt<'mir, 'tcx> for MiriEvalContext<'mir, 'tcx> {}
pub trait EvalContextExt<'mir, 'tcx: 'mir>: MiriEvalContextExt<'mir, 'tcx> {
```

bjorn3 commented on src/data_race.rs:
```diff
@@ -0,0 +1,1407 @@
+        let data_race = &*this.memory.extra.data_race;
+        let old = data_race.multi_threaded.get();
+
+        data_race.multi_threaded.set(false);
```
I believe you can use `let old = data_race.multi_threaded replace(false);`.

bjorn3 commented on src/data_race.rs:
```diff
@@ -0,0 +1,1407 @@
+            "Invalid alt (>=):\n l: {:?}\n r: {:?}",l,r
+        );
+    }
+}
```
nit: missing trailing newline here and in other files.

RalfJung commented:
@JCTyblaidd thanks a lot! This is coming quite out of the blue, but a welcome surprise. :)

Unfortunately I will probably not have the time during the week to review such a massive PR. I hope next week-end will work, but it might also be 2 weeks until I have a few free hours to do the review.

> Data-race detection is only enabled after the first thread is created, so should have minimal overhead for non-concurrent execution.

Conceivably a program could briefly create a thread for a one-off task and then continue running mostly sequentially afterwards. So I think it would still be good to do some benchmarking to measure the impact of this. Unfortunately Miri does not have an established benchmarking infrastructure, all we have are [two little crates](https://github.com/rust-lang/miri/tree/master/bench-cargo-miri) that I used for some ad-hoc benchmarking in the past. Could you measure the impact of your PR on those cases (after adding a dummy thread created in the beginning)? `hyperfine` would be a good way to get some reliable numbers. (There's also the `benches` folder which exists from ye olde tiems and which I never used for anything...)

bjorn3 commented:
Maybe it could be disabled again once all threads except the main thread have been joined.

RalfJung commented:
Well, then there could be a background thread just sitting there waiting for some cleanup to need happening, but not actually doing anything.

Basically, I am worried about making "there's a thread somewhere doing nothing" becoming a huge performance cliff.

JCTyblaidd commented:
Only did the tests for one of the benchmarks, this only adds one dangling thread so the overhead might be reduced by being inside the small vector size limit. Will try get more numbers later.

----

##### Single-threaded:

time ./miri run ./bench-cargo-miri/mse/src/main.rs -Zmiri-disable-stacked-borrows
real	0m1.268s
user	0m1.053s
sys	0m0.232s

---

```
hyperfine --warmup 2 'target/x86_64-unknown-linux-gnu/release/miri --sysroot ~/.cache/miri/HOST ./bench-cargo-miri/mse/src/main.rs -Zmiri-disable-stacked-borrows'
Benchmark #1: target/x86_64-unknown-linux-gnu/release/miri --sysroot ~/.cache/miri/HOST ./bench-cargo-miri/mse/src/main.rs -Zmiri-disable-stacked-borrows
  Time (mean ± σ):     716.6 ms ±   6.6 ms    [User: 694.5 ms, System: 20.4 ms]
  Range (min … max):   708.0 ms … 729.2 ms    10 runs
```

---

##### Single-threaded with overhead:

time ./miri run ./bench-cargo-miri/mse_and_dangling_thread/src/main.rs -Zmiri-disable-stacked-borrows
warning: thread support is experimental.


real	0m2.184s
user	0m1.176s
sys	0m0.283s

---
```
hyperfine --warmup 2 'target/x86_64-unknown-linux-gnu/release/miri --sysroot ~/.cache/miri/HOST ./bench-cargo-miri/mse_and_dangling_thread/src/main.rs -Zmiri-disable-stacked-borrows'
Benchmark #1: target/x86_64-unknown-linux-gnu/release/miri --sysroot ~/.cache/miri/HOST ./bench-cargo-miri/mse_and_dangling_thread/src/main.rs -Zmiri-disable-stacked-borrows
  Time (mean ± σ):     744.0 ms ±   7.1 ms    [User: 726.1 ms, System: 17.9 ms]
  Range (min … max):   736.5 ms … 759.7 ms    10 runs
```


---

#####  Changed benchmark with variant:

```
fn main() {
    let thread = std::thread::spawn(|| 4);
    for _ in 0..2 {
        mse(PCM.len(), PCM, EXPECTED);
    }
    assert_eq!(4, thread.join().unwrap());
}
```

RalfJung commented:
Thanks for measuring this! Is the "716.6 ms ±   6.6 ms" using your branch, or master? I am surprised that the difference is so small (doesn't this affect every single memory access?), also given that `time` via `./miri` shows a much bigger impact.

JCTyblaidd commented:
"716.6 ms ± 6.6 ms" is with my branch, which should is similar to current upstream master with the only overhead of an easily speculated branch that is always false to skip data-race checking, seems to be within the margin of error. This could be the benchmark being used though, if the values do not escape into memory and are kept inline? More benchmarks would probably be better but I can't get the other one to compile, it complains about importing serde_json.

# Upstream master
```
  Time (mean ± σ):     723.2 ms ±   4.3 ms    [User: 706.9 ms, System: 16.1 ms]
  Range (min … max):   716.6 ms … 728.7 ms    10 runs
```
---

The overhead might be minimal in the above "744.0 ms ±   7.1 ms" case  because(maybe wrong):
 - values are re-used? the range-map hence needs minimal updating for internal memory cells
 - there are no atomic/join/thread-create operations during the calculation, so the Timestamp remains constant and hence the range-map might be able to remain in the state of WRITE_THREAD=0, LAST_WRITE=K, READS = {0: K} (inline) for the entire allocation with no need to differentiate?
 - only 2 threads so the read vector for data-races remains an inline smallvec
 - as above, values don't escape to memory? 

Probably need more benchmarks to get a clearer picture.

RalfJung commented on src/data_race.rs:
```diff
@@ -0,0 +1,1189 @@
+struct MemoryCellClocks {
+
+    /// The vector-clock of the last write, only one value is stored
+    ///  since all previous writes happened-before the current write
```
I don't think that is entirely true... e.g. if you have two relaxed writes (by two different threads) to the same location, the first write does not happen-before the second. It is "coherence-before", but that does not imply "happens-before".

Storing just one timestamp might still be good enough, but I do not think the argument holds.

RalfJung commented on src/data_race.rs:
```diff
@@ -0,0 +1,1189 @@
+
+    /// The identifier of the thread that performed the last write
+    ///  operation
+    write_index: VectorIdx,
```
We have a `ThreadId` type, why is this `VectorIdx`?

RalfJung commented on src/data_race.rs:
```diff
@@ -0,0 +1,1189 @@
+    ///  operation
+    write_index: VectorIdx,
+
+    /// The vector-clock of the set of previous reads
```
I think there should be a `.` at the end of this line? At least my attempts to parse this entire comment as a single sentence failed.

RalfJung commented on src/data_race.rs:
```diff
@@ -0,0 +1,1189 @@
+    ///  thread last read this value.
+    read: VClock,
+
+    /// Atomic acquire & release sequence tracking clocks
```
```suggestion
    /// Atomic acquire & release sequence tracking clocks.
```

RalfJung commented on src/data_race.rs:
```diff
@@ -0,0 +1,1189 @@
+    /// The clock-vector for the set of atomic read operations
+    ///  used for detecting data-races with non-atomic write
+    ///  operations
+    read_vector: VClock,
```
So what are the times stored in here? Is it when each thread performed the last atomic read?

RalfJung commented on src/data_race.rs:
```diff
@@ -0,0 +1,1189 @@
+///  of a thread, contains the happens-before clock and
+///  additional metadata to model atomic fence operations
+#[derive(Clone, Default, Debug)]
+struct ThreadClockSet {
```
Please move this type up above where it is first used.

RalfJung commented on src/data_race.rs:
```diff
@@ -0,0 +1,1189 @@
+    ///  thread currently, this is set to None
+    ///  after the vector index has been re-used
+    ///  and hence the value will never need to be
+    ///  read during data-race reporting
```
I do not understand this comment. How can it ever be correct to re-use an index? How does that imply that the value "will never need to be read"?

RalfJung commented on src/data_race.rs:
```diff
@@ -0,0 +1,1189 @@
+    /// The name of the thread, updated for better
+    ///  diagnostics when reporting detected data
+    ///  races
+    thread_name: Option<Box<str>>,
```
Isn't this information already present in the concurrency subsystem? Why copy it here?

RalfJung commented on src/data_race.rs:
```diff
@@ -0,0 +1,1189 @@
+    ///  between before and after a thread launch
+    /// Any data-races must be recorded after this
+    ///  so concurrent execution can ignore recording
+    ///  any data-races
```
Any race must be recorded so it can ignore recording? I am confused.

RalfJung commented on src/data_race.rs:
```diff
@@ -0,0 +1,1189 @@
+
+    /// Mapping of a vector index to a known set of thread
+    ///  clocks, this is not directly mapping from a thread id
+    ///  since it may refer to multiple threads
```
Can you add a top-level comment somewhere explaining this entire business about reusing `VectorIdx` and one idx being for multiple threads? It seems very non-trivial.

RalfJung commented on src/data_race.rs:
```diff
@@ -0,0 +1,1189 @@
+    vector_info: RefCell<IndexVec<VectorIdx, ThreadId>>,
+
+    /// The mapping of a given thread to assocaited thread metadata
+    thread_info: RefCell<IndexVec<ThreadId, ThreadExtraState>>,
```
Wouldn't it be better to add that metadata in `thread.rs` as a new field, than have a separate map here?

RalfJung commented on src/data_race.rs:
```diff
@@ -0,0 +1,1189 @@
+
+    /// Detect data-races for an unsychronized read operation, will not perform
+    ///  data-race threads if `multi-threaded` is false, either due to no threads
+    ///  being created or if it is temporarily disabled during a racy read or write
```
s/racy/atomic? "racy" doesn't seem right.^^ We don't want to allow races, after all.

RalfJung commented on src/data_race.rs:
```diff
@@ -0,0 +1,1189 @@
+                ) {
+                    ("ATOMIC_LOAD", idx, &atomic.read_vector)
+                }else{
+                    unreachable!("Failed to find report data-race for non-atomic operation: no race found")
```
"find report"? One of these verbs is incorrect I guess?

RalfJung commented on src/data_race.rs:
```diff
@@ -0,0 +1,1189 @@
+            }).map(|idx| VectorIdx::new(idx))
+    }
+
+    /// Report a data-race found in the program
```
```suggestion
    /// Report a data-race found in the program.
```

RalfJung commented on src/data_race.rs:
```diff
@@ -0,0 +1,1189 @@
+    }
+
+    /// Report a data-race found in the program
+    ///  this finds the two racing threads and the type
+    ///  of data-race that occured, this will also
```
```suggestion
    /// This finds the two racing threads and the type
    /// of data-race that occurred. This will also
```

RalfJung commented on src/data_race.rs:
```diff
@@ -0,0 +1,1189 @@
+pub struct VClockAlloc {
+
+    /// Range of Vector clocks, mapping to the vector-clock
+    ///  index of the last write to the bytes in this allocation
```
"mapping to ... to ..."? I cannot entirely parse this.

RalfJung commented on src/data_race.rs:
```diff
@@ -0,0 +1,1189 @@
+    /// The vector-clock of the set of previous reads
+    ///  each index is set to the timestamp that the associated
+    ///  thread last read this value.
+    read: VClock,
```
If I understand the code correctly, this is reset to 0 on each write. That should also be said here.

RalfJung commented on src/data_race.rs:
```diff
@@ -0,0 +1,1189 @@
+    /// Update memory cell data-race tracking for atomic
+    ///  load acquire semantics, is a no-op if this memory was
+    ///  not used previously as atomic memory
+    fn acquire(&mut self, clocks: &mut ThreadClockSet, index: VectorIdx) -> Result<(), DataRace> {
```
Could you rename this to `load_acquire`, for consistency?

RalfJung commented on src/data_race.rs:
```diff
@@ -0,0 +1,1189 @@
+
+    /// Update the memory cell data-race tracking for atomic
+    ///  store release semantics
+    fn release(&mut self, clocks: &ThreadClockSet, index: VectorIdx) -> Result<(), DataRace> {
```
Could you rename this to `store_release`, for consistency?

RalfJung commented on src/data_race.rs:
```diff
@@ -0,0 +1,1189 @@
+    ///  sequence exists in the memory cell, required
+    ///  since read-modify-write operations do not
+    ///  invalidate existing release sequences 
+    release_sequences: VSmallClockSet,
```
I am somewhat surprised that release sequences need special treatment here... I seem to recall having seen a "clock-based model" where they just fall out without special treatment. But I might misremember. I'll have to dig deeper into that paper you are implementing it seems.

RalfJung reviewed (commented):
This looks really great overall, thanks a lot for this impressive piece of work! My comment are mostly formatting and grammar nits, and some naming suggestions.

I did not fully understand every part of the algorithm, but I do not necessarily have to. What I saw mostly makes intuitive sense, except for release sequences which I need to look into further, and some parts of the handling of non-atomic accesses.

What I missed is anything specific to `SeqCst`. `SeqCst` induces a global ordering on all sequentially consistent operations, shouldn't that affect the data race detector? Currently, if I saw correctly, `SeqCst` is fully equivalent to always using the strongest available release/acquire ordering.

JCTyblaidd commented on src/thread.rs:
```diff
@@ -501,6 +508,7 @@ impl<'mir, 'tcx: 'mir> ThreadManager<'mir, 'tcx> {
             if thread.state == ThreadState::Enabled {
                 if !self.yield_active_thread || id != self.active_thread {
                     self.active_thread = id;
+                    data_race.thread_set_active(self.active_thread);
```
The AllocExtra functions for read, write and deallocate do not have access to the current active thread in its current location, and need the current id to perform the validation, and also the current thread name for reporting diagnostics. So either a local copy needs to be kept, or the associated thread metadata should be moved to a Rc<RefCell<..>> so that each AllocExtra for the data-race detector can access the value reference, (this will add the overhead for looking up a thread-id to a vector-idx for each access). Should I move to the Rc<RefCell<..>> model?

RalfJung replied:
Ah, so the problem is that we do not have access to the `MemoryExtra` here? Yeah, that's annoying...

Could you leave a comment explaining the situation? I have some ideas for rearranging these access hooks so that they *can* access `MemoryExtra` When/if that materializes, I'll ping you and maybe you can clean this up then. :)

JCTyblaidd replied:
The currently active thread is stored in the ```Evaluator<'mir,'tcx>{ ..., threads: ThreadManager<..> .. }``` instead of ```MemoryExtra```, but the same issue occurs in either location, i'll add a comment now.

RalfJung replied:
Ah, fair -- but moving it to `MemoryExtra` would be easy (whether we put something into `Evaluator` or `MemoryExtra` is pretty arbitrary). So once `MemoryExtra` is available in these hooks, the problem can be solved cleanly.

JCTyblaidd commented:
I've pushed a few commits that should fix the requests. As i added into a comment inside the data-race descriptor code, per the C++ specification on sequentially consistent ordering:

```A load operation with this memory order performs an acquire operation, a store performs a release operation, and read-modify-write performs both an acquire operation and a release operation, plus a single total order exists in which all threads observe all modifications in the same order (see Sequentially-consistent ordering below) ```

This means, since we currently do not model weak memory effects and the scheduling algorithm already provides a global total order of all memory operations, that seq-cst loads/stores are equivalent to acquire loads/release stores and a seq-cst RMW is equivalent to a acquire-release RMW.

JCTyblaidd commented:
And regarding vector index re-use, i tidied up the explanation in the comments. But the high-level view is a vector-index can only be re-used once the thread has terminated, been joined, and the effects of the join happen-before the current vector clock of all currently active threads.

Therefore no data-races can possible be reported with the old thread, since the thread local happens-before clocks do not decrease, any stored timestamps from that index must be less than all active vector clocks, which means that no data-races can ever be reported at that index, so the vector-index stops providing any value. The program will then advance the vector index by 1 and then allow the vector-index to be re-used for a newly created thread.

RalfJung commented on src/machine.rs:
```diff
@@ -144,8 +148,14 @@ impl MemoryExtra {
         };
+        let data_race = if config.data_race_detector {
+            Some(Rc::new(data_race::GlobalState::new()))
+        }else{
```
Please Ctrl-F for `}else` and make sure to add spaces (this still occurs a few times).

RalfJung commented on src/shims/posix/linux/sync.rs:
```diff
@@ -78,7 +78,17 @@ pub fn futex<'tcx>(
+            //   atomically and totally ordered with respect to other futex
+            //   operations on the same futex word."
+            // SeqCst is total order over all operations, so uses acquire,
+            // either are equal under the current implementation.
```
Even "acquire" doesn't totally order *reads* on a single location though... so this sounds like it should be `SeqCst`.

RalfJung commented on src/sync.rs:
```diff
@@ -62,7 +62,7 @@ struct Mutex {
     queue: VecDeque<ThreadId>,
     /// Data race handle
-    data_race: DataRaceLockHandle
+    data_race: VClock
```
In the other file you have these nice comments explaining what these clocks represent. Can you do the same here (also for the other things in this file)? I think it'd be something like "the clock of the thread that last released the mutex", or so?

RalfJung commented on src/vector_clock.rs:
```diff
@@ -0,0 +1,660 @@
+
+/// A vector clock index, this is associated with a thread id
+/// but in some cases one vector index may be shared with
+/// multiple thread ids id it safe to do so.
```
```suggestion
/// multiple thread ids if it safe to do so.
```

RalfJung commented:
(I only had time to go over the first of your new commits today, will continue tomorrow. Thanks for your patience!)

RalfJung commented:
> But the high-level view is a vector-index can only be re-used once the thread has terminated, been joined, and the effects of the join happen-before the current vector clock of all currently active threads.

Ah, with this extra last condition about all threads, that makes of sense -- thanks!

RalfJung commented on src/data_race.rs:
```diff
@@ -0,0 +1,1369 @@
+        Ok(old)
+    }
+
+    /// Perform an atomic compare and exchange at a given memory location
+    /// on success an atomic RMW operation is performed and on failure
```
```suggestion
    /// Perform an atomic compare and exchange at a given memory location.
    /// On success an atomic RMW operation is performed and on failure
```

RalfJung commented on src/shims/posix/sync.rs:
```diff
@@ -62,7 +62,10 @@ fn mutex_get_kind<'mir, 'tcx: 'mir>(
-    ecx.read_scalar_at_offset(mutex_op, offset, ecx.machine.layouts.i32)
+    ecx.read_scalar_at_offset_atomic(
+        mutex_op, offset, ecx.machine.layouts.i32,
+        AtomicReadOp::Acquire
```
You added extra vector clocks to model mutex synchronization... so why should these accesses here be release/acquire?  Isn't it it enough to do relaxed accesses? In fact, what goes wrong when using non-atomic accesses? (I imagine something could go wrong but it is not immediately obvious to me.)

RalfJung commented on src/shims/posix/thread.rs:
```diff
@@ -15,25 +15,33 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         this.tcx.sess.warn(
-            "thread support is experimental. \
-             For example, Miri does not detect data races yet.",
+            "thread support is experimental, no weak memory effects are currently emulated.",
```
```suggestion
            "thread support is experimental and incomplete: weak memory effects are not emulated.",
```

RalfJung commented on src/data_race.rs:
```diff
@@ -0,0 +1,1369 @@
+    /// any subsequent calls to `validate_lock_acquire` as well
+    /// as any previous calls to this function after any
+    /// `validate_lock_release` calls.
+    pub fn validate_lock_release_shared(&self, lock: &mut VClock, thread: ThreadId) {
```
So, the difference between "validate_lock_release" and "validate_lock_release_shared" is that the former *overwrites* the old clock, but the latter joins them.

For a normal lock, shouldn't it be the case that both are equivalent? Since the thread that released the lock previously acquired it, its clock will be ahead of the lock clock, so `join` and `clone_from` do the same? (I am trying to figure out why both of these functions exist.)

Also, I noticed "validate_lock_release" is used for condvars, not just for locks. That makes the naming a bit confusing, though I do not have a better proposal. Is this use for condvars related to using `clone_from` vs `join`?

RalfJung commented on src/data_race.rs:
```diff
@@ -0,0 +1,1369 @@
+        let (index, mut clocks) = self.load_thread_state_mut(thread);
+        clocks.increment_clock(index);
+        clocks.clock.join(&lock);
+        clocks.increment_clock(index);
```
Is there an easy explanation for why the clock needs to be advanced *twice*?

RalfJung commented on tests/compile-fail/data_race/atomic_read_write_race_alt.rs:
```diff
@@ -0,0 +1,31 @@
+
+use std::thread::spawn;
+use std::sync::atomic::AtomicUsize;
+use std::sync::atomic::Ordering;
```
Maybe "atomic_read_na_write" would be clearer as filename, to indicate that the write is *not* atomic. (Same for the other tests.) The "race" suffix is redundant as the folder name already says as much.

When there are multiple tests for the same thing I usually numbered them 1, 2, ...; that seems more scalable than "_alt" suffixes.

RalfJung commented on tests/compile-fail/data_race/dangling_thread_async_race.rs:
```diff
@@ -0,0 +1,44 @@
+    // Spawn and immediately join a thread
+    //  to execute the join code-path
+    //  and ensure that data-race detection
+    //  remains enabled
```
```suggestion
    //  remains enabled nevertheless.
```

RalfJung commented on tests/compile-fail/data_race/dangling_thread_race.rs:
```diff
@@ -0,0 +1,41 @@
+    // Spawn and immediately join a thread
+    //  to execute the join code-path
+    //  and ensure that data-race detection
+    //  remains enabled
```
```suggestion
    //  remains enabled nevertheless.
```

RalfJung commented on tests/compile-fail/data_race/enable_after_join_to_main.rs:
```diff
@@ -0,0 +1,38 @@
+unsafe impl<T> Sync for EvilSend<T> {}
+
+pub fn main() {
+    // Enable and the join with multiple threads
```
```suggestion
    // Enable and then join with multiple threads.
```

RalfJung commented on tests/compile-fail/data_race/relax_acquire_race.rs:
```diff
@@ -0,0 +1,43 @@
+
+        let j3 = spawn(move || {
+            if SYNC.load(Ordering::Acquire) == 2 {
+                *c.0 //~ ERROR Data race
```
Isn't this test very scheduler-dependent? And to make them robust the threads would have to `loop` if they `SYNC.load` the wrong thing?

If so, no need to change the code, but please leave a comment so that if/when we change the scheduler, we have an idea for what to do here.

RalfJung commented on tests/compile-fail/data_race/release_seq_race.rs:
```diff
@@ -0,0 +1,47 @@
+        let j3 = spawn(move || {
+            sleep(Duration::from_millis(1000));
+            if SYNC.load(Ordering::Acquire) == 3 {
+                *c.0 //~ ERROR Data race
```
Oh this is a nasty test, I like it. :D

Same comments wrt. scheduler-dependence as before applies, though.

RalfJung commented on tests/compile-fail/data_race/rmw_race.rs:
```diff
@@ -0,0 +1,44 @@
+
+        let j3 = spawn(move || {
+            if SYNC.load(Ordering::Acquire) == 3 {
+                *c.0 //~ ERROR Data race
```
Same scheduler-dependence issue. (And another very nice test. :D )

RalfJung commented on tests/run-pass/concurrency/disable_data_race_detector.rs:
```diff
@@ -0,0 +1,28 @@
+        });
+
+        let j2 = spawn(move || {
+            *c.0 = 64; //~ ERROR Data race
```
```suggestion
            *c.0 = 64; //~ ERROR Data race (but not detected as the detector is disabled)
```

RalfJung reviewed (commented):
Thanks a lot! This is my final round of comments/questions.

Sadly, I do not have the time to actually understand the underlying data race detection algorithm. But thankfully you based this of a POPL paper, so I am confident that if necessary, I can dig into that paper to understand the grand scheme of things. If there are ways in which you diverged from the paper, or particular implementation choices you made (beyond what you already explained in comments), I'd much appreciate if you could add those to the comment at the top of the data race detector, so that whoever tries to match up paper and implementation in the future will have it as easy as you can make it.

I also like your compile-fail tests. Are there others you could imagine having that you didn't have the time to write? If so, please open an issue, so maybe someone can pick this up in the future. One thing I wondered about is a test related to release sequences -- something that is UB because the release sequence got broken.

JCTyblaidd commented:
Fixed review changes, I added some comments to `validate_lock_release(_shared)`, incrementing the clocks twice per operation seems to be unnecessary and I think was a hold-over from an earlier version, so changed to a single increment.  mutex/rwlock/condvar get_set_id have been made relaxed, they were converted to atomic since the current internal implementation reports a data-race when non-atomic accesses are used(for mutex is in pthread_mutex_lock, i think from mutex_get_or_create_id), it might be a bug in the pthreads shim implemetation, I am not sure either way, if it is a bug then the access should be converted back to non-atomic once fixed.

RalfJung commented:
I am not surprised to see atomic accesses in the mutex implementation, this makes sense -- they just should not induce any synchronization themselves. So `Relaxed` is just right, thanks!

RalfJung commented on src/data_race.rs:
```diff
@@ -37,6 +37,24 @@
 //!
+//! The timestamps used in the data-race detector assign each sequence of non-atomic operations
+//! followed by a single atomic or concurrent operation a single timestamp.
+//! Write, Read, Write, ThreadJoin will be represented by a single timestamp value on a thread
```
Is there a full stop missing at the end here?

Also, does this match what the paper does? If the paper increments before and after, then adding a comment that this is somewhat strange but see the paper for details seems better than trying to adjust the algorithm the paper authors came up with.

RalfJung commented on src/data_race.rs:
```diff
@@ -499,7 +517,7 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: MiriEvalContextExt<'mir, 'tcx> {
     }
 
     /// Perform an atomic compare and exchange at a given memory location
```
```suggestion
    /// Perform an atomic compare and exchange at a given memory location.
```

RalfJung commented on src/data_race.rs:
```diff
@@ -1321,9 +1335,11 @@ impl GlobalState {
     /// any subsequent calls to `validate_lock_acquire` as well
     /// as any previous calls to this function after any
     /// `validate_lock_release` calls.
+    /// For normal locks this should be equivalent to `validate_lock_release`
+    /// this function only exists for joining over the set of concurrent readers
```
```suggestion
    /// For normal locks this should be equivalent to `validate_lock_release`.
    /// This function only exists for joining over the set of concurrent readers
```

RalfJung commented on src/shims/posix/sync.rs:
```diff
@@ -62,9 +62,11 @@ fn mutex_get_kind<'mir, 'tcx: 'mir>(
 ) -> InterpResult<'tcx, ScalarMaybeUninit<Tag>> {
     let offset = if ecx.pointer_size().bytes() == 8 { 16 } else { 12 };
+    //FIXME: this has been made atomic to fix data-race reporting inside the internal
+    // mutex implementation, it may not need to be atomic.
```
No need to leave a FIXME here I think (same in the rest of this file).

RalfJung commented on src/data_race.rs:
```diff
@@ -0,0 +1,1385 @@
+/// Vector clock metadata for a logical memory allocation.
+#[derive(Debug, Clone)]
+pub struct VClockAlloc {
+    /// Range of Vector clocks, this gives each byte a potentially
+    /// unqiue set of vector clocks, but merges identical information
+    /// together for improved efficiency.
```
```suggestion
    /// Assigning each byte a MemoryCellClocks.
```
"Set" is a bit confusing here since it sounds like it might be a `HashSet` or so. Also, the fact that `RangeMap` is internally more efficient should not matter at this level.

RalfJung commented on src/data_race.rs:
```diff
@@ -39,21 +39,14 @@
-//! Both result in the sequence on thread x up to and including the atomic release as happening
-//! before the acquire.
+//! As per the paper a threads timestamp is only incremented after a release operation is performed
+//! so some atomic operations that only perform acquires do not increment the timestamp, due to shared
```
```suggestion
//! so some atomic operations that only perform acquires do not increment the timestamp. Due to shared
```

RalfJung commented on src/data_race.rs:
```diff
@@ -640,7 +633,9 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: MiriEvalContextExt<'mir, 'tcx> {
                 }
-                Ok(())
+                
+                // Increment timestamp if hase release semantics
```
```suggestion
                // Increment timestamp on case of release semantics
```

RalfJung commented:
Thanks again @JCTyblaidd for the PR and for working with me through the review. This is an amazing addition to Miri. :-)

@bors r+

bors commented:
:pushpin: Commit cbb695f782dceca959661e9c57d7aeb120cbc1d8 has been approved by `RalfJung`

<!-- @bors r=RalfJung cbb695f782dceca959661e9c57d7aeb120cbc1d8 -->
<!-- homu: {"type":"Approved","sha":"cbb695f782dceca959661e9c57d7aeb120cbc1d8","approver":"RalfJung"} -->

bors commented:
:hourglass: Testing commit cbb695f782dceca959661e9c57d7aeb120cbc1d8 with merge d473242ecdffe9f06469168a857e1f54e75c0090...
<!-- homu: {"type":"BuildStarted","head_sha":"cbb695f782dceca959661e9c57d7aeb120cbc1d8","merge_sha":"d473242ecdffe9f06469168a857e1f54e75c0090"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/miri/runs/1470661218)
Approved by: RalfJung
Pushing d473242ecdffe9f06469168a857e1f54e75c0090 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"RalfJung","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/miri/runs/1470661218"},"merge_sha":"d473242ecdffe9f06469168a857e1f54e75c0090"} -->
